third_party/move/move-compiler-v2/src/pipeline/unused_assignment_checker.rs

```diff
@@ -0,0 +1,88 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! Implements a pipeline that checks and gives warning on unused assignments.
+//! Prerequisite: live variable annotation.
+//! Side effect: the `LiveVarAnnotation` will be added to the function target annotations.
+
+use crate::pipeline::livevar_analysis_processor::LiveVarAnnotation;
+use codespan_reporting::diagnostic::Severity;
+use move_binary_format::file_format::CodeOffset;
+use move_model::{ast::TempIndex, model::FunctionEnv};
+use move_stackless_bytecode::{
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{AttrId, Bytecode},
+};
+
+pub struct UnusedAssignmentChecker {}
+
+impl UnusedAssignmentChecker {
+    /// Check if the assignment to `dst` at offset after the position given by `offset` and `after`.
+    fn check_unused_assignment(
+        target: &FunctionTarget,
+        id: AttrId,
+        offset: CodeOffset,
+        dst: TempIndex,
+        after: bool,
+    ) {
+        let data = target.data;
+        // only check for user defined variables
+        if let Some(dst_name) = data.local_names.get(&dst) {
```

Comment:
Yes. Done.

Reply:
Which test is it? I seem to not be able to find such a test.

Reply:
Any answer?

Reply:
Sorry, I missed this comment. See for example `third_party/move/move-compiler-v2/tests/unused-assignment/unused_call_assign_shadow.exp`. The flow-insensitive checker warns unused parameters; the unused assignment checker does NOT warn them, because the parameters are always inferred to be alive before the first instruction by the current analysis, even if they are not used anywhere at all.

Reply:
@fEst1ck I still don't see it, and maybe I was not clear about what I was asking, I have tried to make it concrete below:

We need to have a function `foo` with a parameter `x`. We then assign a value to `x` in the function `foo`. After this assignment, there should be no read of `x` (thus, there is an unused assignment to a parameter). I do not think the test case you point to (or others, at the time I inspected them) has a similar case.

Reply:
@vineethk Thanks for the clarification. I added a new test `third_party/move/move-compiler-v2/tests/unused-assignment/unused_assign_to_param.move` for this.